relayer/spy_relayer/src/backends/default/listener.ts

```diff
@@ -0,0 +1,272 @@
+      return "Wrong payload type";
+    }
+    try {
+      parsedPayload = await this.parsePayload(parsedVaa.payload);
+    } catch (e: any) {
+      return "Payload parsing failure";
+    }
+
+    // Verify we want to relay this request
+    if (
+      !this.verifyIsApprovedToken(parsedPayload) ||
+      !this.verifyFeeSpecified(parsedPayload)
+    ) {
+      return "Validation failed";
+    }
+
+    // Great success!
+    return { ...parsedVaa, payload: parsedPayload };
+  }
+
+  /** Get spy filters for all emitters we care about */
+  public async getEmitterFilters(): Promise<TypedFilter[]> {
+    let filters: {
+      emitterFilter: { chainId: ChainId; emitterAddress: string };
+    }[] = [];
+    for (let i = 0; i < this.env.spyServiceFilters.length; i++) {
+      this.logger.info("Getting spyServiceFiltera " + i);
+      const filter = this.env.spyServiceFilters[i];
+      this.logger.info(
+        "Getting spyServiceFilter[" +
+          i +
+          "]: chainId = " +
+          filter.chainId +
+          ", emmitterAddress = [" +
+          filter.emitterAddress +
+          "]"
+      );
+      const typedFilter = {
+        emitterFilter: {
+          chainId: filter.chainId as ChainId,
+          emitterAddress: await encodeEmitterAddress(
+            filter.chainId,
+            filter.emitterAddress
+          ),
+        },
+      };
+      this.logger.info("Getting spyServiceFilterc " + i);
+      this.logger.info(
+        "adding filter: chainId: [" +
+          typedFilter.emitterFilter.chainId +
+          "], emitterAddress: [" +
+          typedFilter.emitterFilter.emitterAddress +
+          "]"
+      );
+      this.logger.info("Getting spyServiceFilterd " + i);
+      filters.push(typedFilter);
+      this.logger.info("Getting spyServiceFiltere " + i);
+    }
+    return filters;
+  }
+
+  /** Process and validate incoming VAAs from the spy. */
+  public async process(rawVaa: Uint8Array): Promise<void> {
+    const validationResults: ParsedVaa<ParsedTransferPayload> | string =
+      await this.validate(rawVaa);
+
+    if (typeof validationResults === "string") {
+      this.logger.debug(`Skipping spied request: ${validationResults}`);
+      return;
+    }
+
+    const parsedVaa: ParsedVaa<ParsedTransferPayload> = validationResults;
+
+    const originChain = parsedVaa.payload.originChain;
+    const originAddress = parsedVaa.payload.originAddress;
+
+    let originAddressNative: string;
+    try {
+      originAddressNative = tryHexToNativeString(originAddress, originChain);
+    } catch (e: any) {
+      this.logger.error(
+        `Failure to convert address "${originAddress}" on chain "${originChain}" to the native address`
+      );
+      return;
+    }
+
+    const key = getKey(parsedVaa.payload.originChain, originAddressNative);
+
+    const isQueued = await checkQueue(key);
+    if (isQueued) {
+      this.logger.error(`Not storing in redis: ${isQueued}`);
+      return;
+    }
```

Comment:
This should likely be turned into a generic function outside of this class, but the `checkQueue()` logic needs to work better first.